File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/youtube/YoutubeTrackFormat.java
package com.sedmelluq.discord.lavaplayer.source.youtube;

import org.apache.http.entity.ContentType;

import java.net.URI;
import java.net.URISyntaxException;

/**
 * Describes an available media format for a track
 */
public class YoutubeTrackFormat {
  private final YoutubeFormatInfo info;
  private final ContentType type;
  private final long bitrate;
  private final long contentLength;
  private final String url;
  private final String signature;

  /**
   * @param type Mime type of the format
   * @param bitrate Bitrate of the format
   * @param contentLength Length in bytes of the media
   * @param url Base URL for the playback of this format
   * @param signature Cipher signature for this format
   */
  public YoutubeTrackFormat(ContentType type, long bitrate, long contentLength, String url, String signature) {
    this.info = YoutubeFormatInfo.get(type);
    this.type = type;
    this.bitrate = bitrate;
    this.contentLength = contentLength;
    this.url = url;
    this.signature = signature;
  }

  /**
   * @return Format container and codec info
   */
  public YoutubeFormatInfo getInfo() {
    return info;
  }

  /**
   * @return Mime type of the format
   */
  public ContentType getType() {
    return type;
  }

  /**
   * @return Bitrate of the format
   */
  public long getBitrate() {
    return bitrate;
  }

  /**
   * @return Base URL for the playback of this format
   */
  public URI getUrl() {
    try {
      return new URI(url);
    } catch (URISyntaxException e) {
      throw new RuntimeException(e);
    }
  }

  /**
   * @return Length in bytes of the media
   */
  public long getContentLength() {
    return contentLength;
  }

  /**
   * @return Cipher signature for this format
   */
  public String getSignature() {
    return signature;
  }
}

File: main/src/main/java/lavaplayer/source/youtube/YoutubeTrackFormat.kt
package lavaplayer.source.youtube

import org.apache.http.entity.ContentType

import java.net.URI
import java.net.URISyntaxException

/**
 * Describes an available media format for a track
 *
 * @param type          Mime type of the format
 * @param bitrate       Bitrate of the format
 * @param contentLength Length in bytes of the media
 * @param baseUrl           Base URL for the playback of this format
 * @param signature     Cipher signature for this format
 * @param signatureKey  The key to use for deciphered signature in the final playback URL
 */
data class YoutubeTrackFormat(
    /**
     * Mime type of the format
     */
    val type: ContentType,
    /**
     * @return Bitrate of the format
     */
    val bitrate: Long,
    /**
     * Length in bytes of the media
     */
    val contentLength: Long,
    private val baseUrl: String,
    /**
     * Cipher signature for this format
     */
    val signature: String?,
    /**
     * The key to use for deciphered signature in the final playback URL
     */
    val signatureKey: String
) {
    /**
     * Format container and codec info
     */
    val info = YoutubeFormatInfo.get(type)

    /**
     * Base URL for the playback of this format
     */
    val url: URI
        get() = try {
            URI(baseUrl)
        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        }
}
